packages/parser/src/index.ts

```diff
@@ -129,25 +147,68 @@ export async function validate<S extends APIDocument = APIDocument>(
 
   // Validate the API against the OpenAPI or Swagger JSON schema definition.
   // NOTE: This is safe to do, because we haven't dereferenced circular $refs yet
-  validateSchema(parser.schema, {
-    colorizeErrors: options?.validate && 'colorizeErrors' in options.validate ? options.validate.colorizeErrors : false,
-  });
+  result = validateSchema(parser.schema, options);
+  if (!result.valid) {
+    return result;
+  }
 
   if (parser.$refs?.circular) {
     if (circular$RefOption === true) {
       // The API has circular reference so we need to do a second pass to fully dereference it.
       dereferenceInternal<S>(parser, parserOptions);
     } else if (circular$RefOption === false) {
       // The API has circular references but we're configured to not permit that.
-      throw new ReferenceError('The API contains circular references');
+      throw new ReferenceError(
+        'The API contains circular references but the validator is configured to not permit them.',
+      );
     }
   }
 
   // Validate the API against the OpenAPI or Swagger specification.
-  validateSpec(parser.schema);
+  const rules = options?.validate?.rules?.openapi;
+  result = validateSpec(parser.schema, {
+    openapi: {
+      'array-without-items': rules?.['array-without-items'] || 'error',
+      'duplicate-non-request-body-parameters': rules?.['duplicate-non-request-body-parameters'] || 'error',
+      'duplicate-operation-id': rules?.['duplicate-operation-id'] || 'error',
+      'non-optional-path-parameters': rules?.['non-optional-path-parameters'] || 'error',
+      'path-parameters-not-in-parameters': rules?.['path-parameters-not-in-parameters'] || 'error',
+      'path-parameters-not-in-path': rules?.['path-parameters-not-in-path'] || 'error',
+    },
+  });
 
-  // If necessary, repair the schema of any anomalies and quirks.
-  repairSchema(parser.schema, args.path);
+  return result;
+}
 
-  return parser.schema;
+/**
+ * A utility to transform the `ValidationResult` from a `validate()` call to a human-readable
+ * string.
+ *
+ */
+export function compileErrors(result: ValidationResult): string {
+  const specName = !result.specification ? 'API definition' : result.specification;
+  const status = !result.valid ? 'failed' : 'succeeded, but with warnings';
+
+  let message = `${specName} schema validation ${status}.\n`;
+  message += '\n';
+
+  if (result.valid === false) {
+    if (result.errors.length) {
+      message += result.errors.map(err => err.message).join('\n\n');
+    }
+  }
+
+  if (result.warnings.length) {
+    if (result.valid === false && result.errors.length) {
+      message += '\n\nWe have also found some additional warnings:\n\n';
+    }
+
+    message += result.warnings.map(warn => warn.message).join('\n\n');
+  }
+
+  if (result.valid === false && result.additionalErrors > 0) {
+    message += `\n\nPlus an additional ${result.additionalErrors} errors. Please resolve the above and re-run validation to see more.`;
+  }
```

Comment:
i'm wondering if `message` should be a `string[]` and on line 213 below, we just return `message.join('\n\n')`? the whitespace stuff is a little hard to follow atm

Reply:
b2f98cc4